### DISCUSSION TITLE
How to "deref" a static_map_ref

### DISCUSSION THREAD

**Discussion by Gibyeng** [NVIDIA/cuCollections#616] [Q&A] [answered]
When using bulk allocation storage for multiple maps, I have no access to the underlying `static_map`, which has member functions like `.size()` and 'retrieve_all' to get keys and values. I only have map_refs of it. Is there any way to dereference it so that I can access the keys and values?

**Comment by PointKernel** [comment]
Similar to other operations like `insert` or `find`, there is `static_map_ref::for_each` #565 available for use. Otherwise, you can always access map contents via `probing_scheme` and `storage_ref`: https://github.com/NVIDIA/cuCollections/blob/5602381ff980d3d665760e235e23b9daa368a65b/include/cuco/static_map_ref.cuh#L215-L227

**Reply by Gibyeng** [reply]
@PointKernel  Thank you so much for the explanation. Would you recommend to use `static_map_ref::for_each` to insert keys and values to a `device_vector` ? or there is more effficient way to move the content of those maps into one big vector.

**Reply by PointKernel** [reply]
`static_map_ref::for_each` is probably the best option for now.

CC @sleeepyjack the block-wise `retrieve` would be beneficial in this case.

**Reply by Gibyeng** [reply]
@PointKernel Hi, it appears that the `for_each` and `retrieve` device functions return only values associated with a single probe key. However, I need to retrieve all keys and values from a map_ref on the device. Could you please provide a simple example to demonstrate this?

**Reply by PointKernel** [reply]
voila https://godbolt.org/z/vqPErncPc

**Reply by Gibyeng** [reply]
@PointKernel  Thanks for your example, I write a kernel [link](https://godbolt.org/z/MzavdKMdd) which copies the stored pairs into a device vector. But I still have a few questions:

1. How to get the actual size of the map in device by map_ref ? (I know there is a `size` function for map but it seems not work for ref types)
2. In the code, thread `tid` get a item from `*(iter+tid)`, and write it into `output[tid]`. Are they coalesced reads? it seems that *(iter) and *(iter+1) may be not adjacent in the memory. 
3. Is it possible to copy/move the content of the map into a device vector directly, like `thrust::copy` or `std::move`?

**Reply by PointKernel** [reply]
1. Instead of printing or storing, you can check if the slot key equals the empty key sentinel. If it does, do nothing; otherwise, increment the counter. Depending on your case, the counter might need to be atomic.
2. "`tid` reads `*(iter+tid)`" is the definition of coalesced reads. :)
3. Not sure what you mean by "copy/move the content of the map into a device vector directly", your `write_ftor` is already doing exactly that.

**Reply by Gibyeng** [reply]
@PointKernel Thank you for your response.
 I think using 'atomicadd' to count the unempty slot for each hashtable solution absolutely works. But does cuco static_map already do it for me when I insert items? 
I mean using a solution like `cudaMemcpyDeviceToDevice` when I say "copy/move the content of the map into a device vector directly". It is because the `cudaMemcpy` solution always performs better than the `write_ftor` kernel, is it right?

**Reply by PointKernel** [reply, marked as answer]
> does cuco static_map already do it for me when I insert items?

For host APIs, yes. When users insert via device APIs, they are responsible for counting the number of successful insertions on their own.

> It is because the cudaMemcpy solution always performs better than the write_ftor kernel

It's two operations. You need `write_ftor` to store the output to the device storage first then use `cudaMemcpyDeviceToDevice` to copy the device storage back to the host.
